pkg/virt-controller/watch/clone/clone.go

```diff
@@ -0,0 +1,375 @@
+			return fmt.Errorf("error getting VM %s in namespace %s from cache: %v", sourceInfo.Name, vmClone.Namespace, err), doNotReenqueue
+		}
+		if !vmExists {
+			err = ctrl.updateStatus(vmClone, syncInfoType{
+				isCloneFailing: true,
+				failEvent:      SnapshotNotCreated,
+				failReason:     fmt.Sprintf("VirtualMachine %s does not exist in namespace %s", vmClone.Spec.Source.Name, vmClone.Namespace),
+			})
+
+			if err != nil {
+				log.Log.Errorf("updating status when source vm does not exist failed: %v", err)
+			}
+
+			return fmt.Errorf("VM %s in namespace %s does not exist", sourceInfo.Name, vmClone.Namespace), doNotReenqueue
+		}
+		sourceVM := obj.(*k6sv1.VirtualMachine)
+
+		syncInfo = ctrl.syncSourceVM(key, sourceVM, vmClone)
+		syncErr = syncInfo.err
+	default:
+		return fmt.Errorf("clone %s is defined with an unknown source type %s", vmClone.Name, sourceInfo.Kind), doNotReenqueue
+	}
+
+	err = ctrl.updateStatus(vmClone, syncInfo)
+	if err != nil {
+		return fmt.Errorf("error updating status: %v", err), syncInfo.reenqueueInfo
+	}
+
+	if syncErr != nil {
+		return fmt.Errorf("sync error: %v", syncErr), syncInfo.reenqueueInfo
+	}
+
+	return nil, syncInfo.reenqueueInfo
+}
+
+func (ctrl *VMCloneController) syncSourceVM(key string, source *k6sv1.VirtualMachine, vmClone *clonev1alpha1.VirtualMachineClone) syncInfoType {
+	var targetType cloneTargetType
+	if vmClone.Spec.Target != nil {
+		targetType = cloneTargetType(vmClone.Spec.Target.Kind)
+	} else {
+		targetType = defaultType
+	}
+
+	switch targetType {
+	case targetTypeVM:
+		return ctrl.syncSourceVMTargetVM(key, source, vmClone)
+
+	default:
+		return syncInfoType{err: fmt.Errorf("target type is unknown: %s", targetType)}
+	}
+}
+
+func (ctrl *VMCloneController) syncSourceVMTargetVM(key string, source *k6sv1.VirtualMachine, vmClone *clonev1alpha1.VirtualMachineClone) syncInfoType {
+	syncInfo := syncInfoType{}
+	logger := log.Log.Object(vmClone)
+
+	var snapshot *snapshotv1alpha1.VirtualMachineSnapshot
+	targetVMInfo := vmClone.Spec.Target
+
+	switch vmClone.Status.Phase {
+	case clonev1alpha1.PhaseUnset, clonev1alpha1.SnapshotInProgress:
+
+		// Create snapshot
+		if vmClone.Status.SnapshotName == nil {
+			snapshot = generateSnapshot(vmClone, source)
+			logger.Infof("creating snapshot %s for clone %s", snapshot.Name, vmClone.Name)
+
+			snapshot, syncInfo.err = ctrl.client.VirtualMachineSnapshot(snapshot.Namespace).Create(context.Background(), snapshot, v1.CreateOptions{})
+			if syncInfo.err != nil {
+				return addErrorToSyncInfo(syncInfo, fmt.Errorf("failed creating snapshot %s for clone %s: %v", snapshot.Name, vmClone.Name, syncInfo.err))
+			}
+
+			ctrl.logAndRecord(vmClone, SnapshotCreated, fmt.Sprintf("created snapshot %s for clone %s", snapshot.Name, vmClone.Name))
+			syncInfo.snapshotName = snapshot.Name
+
+			return addReenqueueToSyncInfo(syncInfo, fmt.Sprintf("snapshot %s was just created, reenqueuing to let snapshot time to finish", snapshot.Name))
```

Comment:
It appears to me that the controller is watching everything it creates, so why all the requeues still?

Reply:
You're right, it's a leftover.
It's removed now. Thanks for the reminder! :+1: